designs/0034-jacobian.md

```diff
@@ -0,0 +1,320 @@
+
+- `constrain(TransformType, OtherArguments...) -> ReturnType`: Transforms a real unconstrained variable to a constrained space.
+- `unconstrain(TransformType, OtherArguments...) -> ReturnType`: Transforms a real constrained variable back to the unconstrained space.
+- `validate(TransformType, OtherArguments...) -> ReturnType`: Validates a real variable against the constraint, returning 1 if valid and 0 otherwise.
+
+### Using the jacobian Keyword
+
+The `jacobian` keyword is introduced within the Constraint block to facilitate the increment of the log of the absolute determinant of the Jacobian. This keyword behaves similarly to the `target` keyword, allowing users to account for the change of variables in the probability density function. The `jacobian` keyword is available in both the `constraints` block and the `transformed parameters` block.
+
+### Declaring Variables with User-defined Constraints
+
+Variables can be declared with user-defined constraints in the data, transformed data, parameters, transformed parameters, and generated quantities blocks. The syntax for declaring a variable with a user-defined constraint is as follows:
+
+```stan
+type<udc=constraint_name(args...)> variable_name;
+```
+
+Here, udc denotes a user-defined constraint , and constraint_name(args...) specifies the name of the constraint along with any arguments it requires.
+
+### Example Usage
+
+```stan
+constraints {
+ upper_bound {
+   real constrain(real x, real upper_bound) {
+     jacobian += x;
+     return upper_bound - exp(x);
+   }
+   real constrain(vector x, real upper_bound) {
+     jacobian += sum(x);
+     return upper_bound - exp(x);
+   }
+   real unconstrain(real x, real upper_bound) {
+     return log(upper_bound - x);
+   }
+   // Used for data
+   int validate(real x, real upper_bound) {
+     // For most this will be
+     return x < upper_bound;
+   }
+ }
+}
+data {
+ real a;
+ real u_bound;
+ int N;
+}
+
+
+transformed data {
+ // Data will use the validate() function from the user-defined constraint
+ real<udc=upper_bound(u_bound)> a_validated;
+}
+
+
+parameters {
+ // User-defined constraint applied to parameters alpha and b
+ real<udc=upper_bound(u_bound)> alpha;
+ vector<udc=upper_bound(u_bound)> b;
+ // In these cases this is the same as calling
+ // real<upper=u_bound> alpha;
+ // vector<upper=u_bound>[N] b;
+}
+```
+
+In this example, upper_bound is a user-defined constraint that is applied to the data and parameters. The validate() function from upper_bound is used to validate a_validated in the transformed data block, and the transformations defined in upper_bound are applied to alpha and b in the parameters block.
+
+### Error Messages
+
+If the `constrain` function of a constraint does not increment `jacobian` a warning is thrown to inform users. In these cases it is better to use a standard function as a new constraint type is most likely not necessary.
```

Comment:
I didn't follow this.  Suppose I want a sum-to-one constraint?  There's no Jacobian adjustment, but I'd like to encapsulate.

I think it's OK to throw a warning here, but sometimes it's justified.

Ideally, we'd be able to tell if variables were constants or autodiff variables.  If the transform is multiplied by another expression, if the expression is constant there's no Jacobian needed.

Reply:
I removed this error and replaced with a few that made more sense.

> Ideally, we'd be able to tell if variables were constants or autodiff variables. If the transform is multiplied by another expression, if the expression is constant there's no Jacobian needed.

Since the `jacobian` will be surrounded by an `if (Jacobian)` it will only get incremented if the user asks for it. We could also make the if statement look like 

```
if (Jacobian && !is_constant<TJacobian>::value) {
...
}
```

In the validator part of the compiler we could look to see if a constrained parameter is multiplied by another expression. What sort of warning would we want to give?